### ISSUE TITLE
Datashader & Altair?

### ISSUE THREAD

**Issue by wmayner** [vega/altair#1752]
I love Altair. The API is beautiful, and in my view, far superior to alternatives like Holoviews. However, the Holoviews ecosystem has one killer feature: datashader. When working with millions of points, Altair is not yet an option, so I'm forced to use Holoviews.

In my ideal world, I could use datashader with Altair.

I figured I'd make this issue to register my dreams with the powers that be :)

**Comment by wmayner** [comment]
Would it really require a full translator? In my head I was thinking of some workaround where datashader produces images which are then inserted into the chart specification. I don't know vega-lite well enough to know whether that makes sense, though. Unfortunately I don't think I would be able to spare the time to write a full translator. (I know it's not the most helpful to get feature requests from people who aren't willing/able to help implement them—sorry!)

**Comment by jakevdp** [comment]
I think a bokeh translator/renderer for vega-lite would be a far easier undertaking than trying to join datashader and vega at the javascript level. They're two entirely different frameworks.

**Comment by jakevdp** [comment]
See also https://github.com/vega/scalable-vega. The project is still young so there are no Python bindings yet, but it's a possible solution to the large dataset issue.

**Comment by wmayner** [comment]
I was thinking of joining them at the Python level. I'm assuming vega-lite can render images; since datashader can produce images, then those images could be embedded in the specification. That would all be before the JS stage, as I'm imagining it.

**Comment by jakevdp** [comment]
Oh, OK. yeah, Vega-Lite can render images, for example: https://vega.github.io/vega-lite/docs/image.html

So you could use Bokeh/Datashader to generate an image, and then display that image in an Altair chart, if that's what you're after. But you're not going to be using the Altair grammar to create the image, and you're not going to have any interactive features short of maybe zooming-in on the image pixels.

**Comment by wmayner** [comment]
Right, it would have to be static. I think that's OK for some purposes. As for the grammar: it might be possible to use the spec as the the input to datashader, though, so Altair's grammar could still be used. That is, the function could take a normal vega-lite spec as input and return a spec where large data has been replaced with a datashaded image. If that's possible, then maybe this lower-barrier workaround would be useful, despite the lack of interactivity, until someone can build a proper translator.

**Comment by jakevdp** [comment]
>  it might be possible to use the spec as the the input to datashader

in other words, create a translator from vega-lite to bokeh/datashader 😀

**Comment by jakevdp** [comment]
Once that heavy-lifting has been done in terms of translating Vega-Lite into something that Bokeh/Datashader understands, I see very little benefit in sending a rendered PNG back so Altair can use it. Why not just visualize it natively in DataShader?

**Comment by joelostblom** [comment]
@jakevdp Do you think it is worthwhile to add a brief section to the docs similar to what plotly has on datashader https://plotly.com/python/datashader/? Just to be explicit about how it can be used and maybe talk a bit about performance with  large data in general with vega and its current limitations.

**Comment by jakevdp** [comment]
Sure - what do you have in mind to put in that section? I'm not aware of any Altair/Datashader interoperability similar to what plotly demonstrates there.

**Comment by joelostblom** [comment]
I was thinking along the lines of what you mentioned in your earlier comments in this thread: showing that the image arrays created from datashader can be displayed with Altair (which I think is mostly what that plotly page does). This would also add some functionality like tooltips, although there is no dynamic rescaling of the image as it is zooming in our out (but that is not on the plotly page either, only holoviews/bokeh has that I believe).

So for the second example on the plotly page:

![image](https://user-images.githubusercontent.com/4560057/115789263-b7813280-a379-11eb-9671-7775c5fa7703.png)

I was thinking of something like this for Altair (the axis needs fixing, not sure if this is the best way):

```python
import altair as alt
import pandas as pd
import numpy as np
import datashader as ds

# Same as for plotly / general datashader
df = pd.read_parquet('https://raw.githubusercontent.com/plotly/datasets/master/2015_flights.parquet')
cvs = ds.Canvas(plot_width=100, plot_height=100)
agg = cvs.points(df, 'SCHEDULED_DEPARTURE', 'DEPARTURE_DELAY')
zero_mask = agg.values == 0
agg.values = np.log10(agg.values, where=np.logical_not(zero_mask))
agg.values[zero_mask] = np.nan

# Convert xarray to df for altair plotting
plot_df = agg.to_dataframe(name='Count_log10').reset_index().dropna()
# alt.data_transformers.enable('data_server')
alt.Chart(plot_df, height=400, width=800).mark_rect().encode(
    alt.X('SCHEDULED_DEPARTURE:O'),
    alt.Y('DEPARTURE_DELAY:O', scale=alt.Scale(reverse=True)),
    alt.Color('Count_log10'),
    alt.Stroke('Count_log10'),
    alt.Tooltip(['SCHEDULED_DEPARTURE', 'DEPARTURE_DELAY', 'Count_log10']))
```

![image](https://user-images.githubusercontent.com/4560057/115789550-2bbbd600-a37a-11eb-94aa-2816f48f4f8e.png)

I haven't looked into the first example in details in case there is any specific plotly  to datashader functionality that can't be done in altair.

**Comment by dangotbanned** [comment]
@joelostblom do you still have plans to add a https://github.com/holoviz/datashader section to the docs, as per (https://github.com/vega/altair/issues/1752#issuecomment-825122086) and (https://github.com/vega/altair/issues/1752#issuecomment-825211031)?

If not, i'm thinking this could either be:
1. Converted to a discussion, to gauge interest in such content
2. Closing the issue as not planned, since there hasn;t been any activity in 2 years

**Comment by joelostblom** [comment]
Let's close it since Vegafusion is the way forward to large datasets in altair.
